ql/src/java/org/apache/hadoop/hive/ql/cache/results/QueryResultsCache.java

```diff
@@ -671,6 +674,15 @@ public void notifyTableChanged(String dbName, String tableName, long updateTime)
    */
   private boolean entryMatches(LookupInfo lookupInfo, CacheEntry entry, Set<CacheEntry> entriesToRemove) {
     QueryInfo queryInfo = entry.getQueryInfo();
+
+    if(!queryInfo.getInputs()
+            .stream()
+            .map(ReadEntity::getTable)
+            .map(Table::getTTable)
+            .map(org.apache.hadoop.hive.metastore.api.Table::getId)
+            .collect(Collectors.toSet()).containsAll(lookupInfo.txnTables))
+        return false;
```

Comment:
Adding a holistic check here incurs some overhead since we duplicate some work that is already done as part of the for loop just below. Since we are already iterating over the read entities it may be better to simplify and move the check inside the loop.

```java
lookupInfo.tableIds.contains(tableUsed.getTTable().getId())
```

Moreover, if we detect that a certain cache entry contains a table ID that is not part of the lookupInfo we should determine if we should/can invalidate and remove that entry from the cache. In the current, approach we simply bail-out and leave the entry inside the cache.